cpp/include/rapidsmpf/streaming/core/message.hpp

```diff
@@ -0,0 +1,127 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <any>
+#include <memory>
+#include <stdexcept>
+#include <typeinfo>
+
+#include <rapidsmpf/error.hpp>
+
+namespace rapidsmpf::streaming {
+
+
+/**
+ * @brief @brief Type-erased message wrapper around a payload.
+ */
+class Message {
+  public:
+    Message() = default;
+
+    /**
+     * @brief Construct a new message from an unique pointer to the payload.
+     *
+     * @tparam T Payload type.
+     * @param payload Non-null unique pointer to the payload.
+     *
+     * @throws std::invalid_argument if @p payload is null.
+     */
+    template <typename T>
+    Message(std::unique_ptr<T> payload) {
+        RAPIDSMPF_EXPECTS(
+            payload != nullptr, "nullptr not allowed", std::invalid_argument
+        );
+        // Conceptually, a `std::unique_ptr` would be sufficient for exclusive ownership.
+        // However, since `std::any` requires its contents to be copyable, we store the
+        // payload in a `std::shared_ptr` instead. This shared ownership is internal only
+        // and never exposed to the user.
+        payload_ = std::shared_ptr<T>(std::move(payload));
+    }
+
+    /** @brief Move construct. @param other Source message. */
+    Message(Message&& other) noexcept = default;
+
+    /** @brief Move assign. @param other Source message. @return *this. */
+    Message& operator=(Message&& other) noexcept = default;
+    Message(Message const&) = delete;
+    Message& operator=(Message const&) = delete;
+
+    /**
+     * @brief Reset the message to empty.
+     */
+    void reset() noexcept {
+        return payload_.reset();
+    }
+
+    /**
+     * @brief Returns true when no payload is stored.
+     *
+     * @return true if empty, false otherwise.
+     */
+    [[nodiscard]] bool empty() const noexcept {
+        return !payload_.has_value();
+    }
+
+    /**
+     * @brief Compare the payload type.
+     *
+     * @tparam T Expected payload type.
+     * @return true if the payload is `typeid(T)`, false otherwise.
+     */
+    template <typename T>
+    [[nodiscard]] bool payload_type() const noexcept {
+        return payload_.type() == typeid(std::shared_ptr<T>);
+    }
```

Comment:
*nit*: This function was renamed from `holds<T>()`. It now doesn't read as well, I think, since `payload_type()` suggests it returns a type, not a bool.

And so `if (msg.payload_type<T>())` doesn't make as much sense to me as a reader as `if (msg.holds<T>())`

Reply:
let's go back to `holds`